docs/topics/coroutines-basics.md

```diff
@@ -1,293 +1,612 @@
+This example demonstrates simple multithreading with coroutines on a shared thread pool.
+
+> Try running the example multiple times. 
+> You may notice that the output order and thread names may change each time you run the program, because the OS decides when threads run.
+> 
+{style="tip"}
+
+## Coroutine scope and structured concurrency
+
+When you run many coroutines in an application, you need a way to manage them as groups.
+Kotlin coroutines rely on a principle called _structured concurrency_ to provide this structure.
+
+According to this principle, coroutines form a tree hierarchy of parent and child tasks with linked lifecycles.
+A coroutine's lifecycle is the sequence of states it goes through from creation until it completes, fails, or is canceled.
+
+A parent coroutine waits for its children to complete before it completes.
+If the parent coroutine fails or is canceled, all its child coroutines are recursively canceled too.
+Keeping coroutines connected this way makes cancellation and error handling predictable and safe.
+
+To maintain structured concurrency, new coroutines can only be launched in a [`CoroutineScope`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-scope/) that defines and manages their lifecycle.
+The `CoroutineScope` includes the _coroutine context_, which defines the dispatcher and other execution properties.
+When you start a coroutine inside another coroutine, it automatically becomes a child of its parent scope.
+The parent coroutine's scope waits for all its children to finish before it completes.
+
+Calling a [coroutine builder function](#coroutine-builder-functions) such as `CoroutineScope.launch()` on a `CoroutineScope` starts a child coroutine of the coroutine associated with that scope.
+Inside the builder's block, the [receiver](lambdas.md#function-literals-with-receiver) is a nested `CoroutineScope`, so any coroutines you launch there become its children.
+
+### Create a coroutine scope with the `coroutineScope()` function
+
+To create a new coroutine scope with the current coroutine context, use the
+[`coroutineScope()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html) function.
+This function creates a root coroutine of the coroutine subtree.
+It's the direct parent of coroutines launched inside the block and the indirect parent of any coroutines they launch.
+`coroutineScope()` executes the suspending block and waits until the block and any coroutines launched in it complete.
+
+Here's an example:
 
 ```kotlin
+// Imports the kotlin.time.Duration to express duration in seconds
+import kotlin.time.Duration.Companion.seconds
+
 import kotlinx.coroutines.*
 
 //sampleStart
-fun main() = runBlocking { // this: CoroutineScope
-    launch { doWorld() }
-    println("Hello")
+suspend fun main() {
+    // Root of the coroutine subtree
+    coroutineScope { // this: CoroutineScope
```

Comment:
The explanation here is good. I'd prefer that every `main()` function written in the guide unconditionally has `withContext()` as its root function, though. Although the guide itself contains a mention of `Dispatchers.Default` being used by default, the code itself doesn't, and if people blindly copy-paste this code, they will have obscure behavior on their hands.
 
Suggestion:
```kotlin
suspend fun main() {
    withContext(Dispatchers.Default) {
        foo()
    }
}

suspend fun foo() {
    // what `main` contains currently
}
```